File: scraper.js
#!/usr/bin/env node

/**
 * TikTok Creative Center Music Scraper
 * Fetches trending music data from TikTok's Creative Center
 */

const puppeteer = require('puppeteer');
const fs = require('fs').promises;
const path = require('path');
const createCsvWriter = require('csv-writer').createObjectCsvWriter;

class TikTokMusicScraper {
  constructor() {
    this.url = 'https://ads.tiktok.com/business/creativecenter/inspiration/popular/music/pc/en';
    this.dataDir = 'data';
  }

  async ensureDataDir() {
    try {
      await fs.mkdir(this.dataDir, { recursive: true });
    } catch (error) {
      // Directory already exists
    }
  }

  async scrapeMusic(maxSongs = 50) {
    console.log('🎵 Starting TikTok Music Scraper...');
    console.log(`📊 Target URL: ${this.url}`);

    const browser = await puppeteer.launch({
      headless: 'new',
      args: ['--no-sandbox', '--disable-setuid-sandbox']
    });

    const page = await browser.newPage();
    const songs = [];

    try {
      console.log('🌐 Launching browser...');
      await page.setViewport({ width: 1920, height: 1080 });

      console.log('📡 Loading TikTok Creative Center...');
      await page.goto(this.url, {
        waitUntil: 'networkidle2',
        timeout: 60000
      });

      console.log('⏳ Waiting for content to load...');
      await page.waitForTimeout(5000);

      // Take screenshot for debugging
      const screenshotPath = path.join(this.dataDir, 'page_screenshot.png');
      await page.screenshot({ path: screenshotPath, fullPage: true });
      console.log(`📸 Screenshot saved to ${screenshotPath}`);

      // Save HTML for analysis
      const htmlPath = path.join(this.dataDir, 'page_source.html');
      const html = await page.content();
      await fs.writeFile(htmlPath, html, 'utf-8');
      console.log(`💾 HTML source saved to ${htmlPath}`);

      // Extract song data - we'll need to inspect the page to find the right selectors
      console.log('🔍 Extracting song data...');

      // Try to find any card/list items
      const songElements = await page.$$('[class*="CardPc"], [class*="card"], [class*="item"], [class*="list"] > div');

      console.log(`Found ${songElements.length} potential song elements`);

      // Extract data from each element
      for (let i = 0; i < Math.min(songElements.length, maxSongs); i++) {
        try {
          const element = songElements[i];
          const text = await element.evaluate(el => el.innerText);

          // Try to extract structured data
          const songData = {
            rank: i + 1,
            raw_text: text,
            scraped_at: new Date().toISOString(),
            title: 'Unknown',
            artist: 'Unknown',
            trend: 'N/A'
          };

          // Try to find specific elements (these selectors may need adjustment)
          try {
            const titleElement = await element.$('[class*="title"], [class*="name"], h3, h4');
            if (titleElement) {
              songData.title = await titleElement.evaluate(el => el.innerText.trim());
            }
          } catch (e) {}

          try {
            const artistElement = await element.$('[class*="artist"], [class*="creator"]');
            if (artistElement) {
              songData.artist = await artistElement.evaluate(el => el.innerText.trim());
            }
          } catch (e) {}

          songs.push(songData);
        } catch (error) {
          console.log(`⚠️  Error extracting song ${i}: ${error.message}`);
        }
      }

      // Also try to extract via page evaluation for more structured data
      const extractedData = await page.evaluate(() => {
        const results = [];

        // Look for common patterns in music listing pages
        const selectors = [
          '[class*="music"]',
          '[class*="song"]',
          '[class*="track"]',
          '[class*="card"]',
          '[role="listitem"]'
        ];

        for (const selector of selectors) {
          const elements = document.querySelectorAll(selector);
          if (elements.length > 0) {
            elements.forEach((el, index) => {
              results.push({
                selector: selector,
                index: index,
                text: el.innerText?.substring(0, 200) || '',
                html: el.innerHTML?.substring(0, 300) || ''
              });
            });
            break; // Use first successful selector
          }
        }

        return results;
      });

      if (extractedData.length > 0) {
        console.log(`✨ Found ${extractedData.length} items using page evaluation`);

        // Merge or replace with better data
        extractedData.forEach((item, index) => {
          if (index < songs.length) {
            songs[index].evaluated_text = item.text;
            songs[index].selector_used = item.selector;
          }
        });
      }

    } catch (error) {
      console.error(`❌ Error during scraping: ${error.message}`);
      console.log('💡 Tip: The page structure may have changed. Check the screenshot and HTML.');
    } finally {
      await browser.close();
    }

    console.log(`✅ Scraped ${songs.length} songs`);
    return songs;
  }

  async saveJSON(data, filename = 'trending_music.json') {
    const filepath = path.join(this.dataDir, filename);
    await fs.writeFile(filepath, JSON.stringify(data, null, 2), 'utf-8');
    console.log(`💾 JSON saved to ${filepath}`);
  }

  async saveCSV(data, filename = 'trending_music.csv') {
    if (!data || data.length === 0) {
      console.log('⚠️  No data to save to CSV');
      return;
    }

    const filepath = path.join(this.dataDir, filename);

    // Get all unique keys
    const keys = new Set();
    data.forEach(item => {
      Object.keys(item).forEach(key => keys.add(key));
    });

    const headers = Array.from(keys).map(key => ({ id: key, title: key }));

    const csvWriter = createCsvWriter({
      path: filepath,
      header: headers
    });

    await csvWriter.writeRecords(data);
    console.log(`📊 CSV saved to ${filepath}`);
  }

  async run(maxSongs = 50) {
    console.log('\n' + '='.repeat(60));
    console.log('🎵 TikTok Creative Center Music Scraper');
    console.log('='.repeat(60) + '\n');

    await this.ensureDataDir();

    const songs = await this.scrapeMusic(maxSongs);

    if (songs.length > 0) {
      await this.saveJSON(songs);
      await this.saveCSV(songs);

      console.log('\n' + '='.repeat(60));
      console.log('📈 Summary');
      console.log('='.repeat(60));
      console.log(`Total songs scraped: ${songs.length}`);
      console.log(`Data directory: ${path.resolve(this.dataDir)}`);
      console.log('\n✨ Scraping completed successfully!');
      console.log('\n💡 Next steps:');
      console.log('   - Check data/page_screenshot.png to see what was captured');
      console.log('   - Review data/page_source.html to find better selectors');
      console.log('   - Update selectors in scraper.js for more accurate data');
    } else {
      console.log('\n⚠️  No data was scraped. Please check the screenshot and HTML source.');
      console.log('💡 You may need to update the selectors in the scraper.');
    }

    return songs;
  }
}

// Run the scraper
if (require.main === module) {
  const scraper = new TikTokMusicScraper();
  scraper.run(50).catch(console.error);
}

module.exports = TikTokMusicScraper;

File: scraper.py
#!/usr/bin/env python3
"""
TikTok Creative Center Music Scraper
Fetches trending music data from TikTok's Creative Center
"""

import json
import csv
import os
from datetime import datetime
from playwright.sync_api import sync_playwright
import time


class TikTokMusicScraper:
    def __init__(self):
        self.url = "https://ads.tiktok.com/business/creativecenter/inspiration/popular/music/pc/en"
        self.data_dir = "data"
        self.ensure_data_dir()

    def ensure_data_dir(self):
        """Create data directory if it doesn't exist"""
        if not os.path.exists(self.data_dir):
            os.makedirs(self.data_dir)

    def scrape_trending_music(self, max_songs=50):
        """
        Scrape trending music data from TikTok Creative Center

        Args:
            max_songs: Maximum number of songs to scrape (default: 50)

        Returns:
            list: List of dictionaries containing song data
        """
        print(f"🎵 Starting TikTok Music Scraper...")
        print(f"📊 Target URL: {self.url}")

        songs_data = []

        with sync_playwright() as p:
            # Launch browser
            print("🌐 Launching browser...")
            browser = p.chromium.launch(headless=True)
            page = browser.new_page()

            # Navigate to the page
            print("📡 Loading TikTok Creative Center...")
            page.goto(self.url, wait_until="networkidle")

            # Wait for content to load
            print("⏳ Waiting for content to load...")
            time.sleep(5)  # Give time for dynamic content

            # Try to find song elements
            # Note: Selectors may need adjustment based on actual page structure
            try:
                # Wait for song list to appear
                page.wait_for_selector('[class*="music"], [class*="song"], [class*="track"]', timeout=10000)

                # Get page content for parsing
                content = page.content()

                # Take a screenshot for debugging
                screenshot_path = f"{self.data_dir}/page_screenshot.png"
                page.screenshot(path=screenshot_path)
                print(f"📸 Screenshot saved to {screenshot_path}")

                # Extract song data (this will need to be customized based on actual HTML structure)
                # For now, we'll create a template structure
                print("🔍 Extracting song data...")

                # This is a placeholder - actual selectors will need to be determined
                # by inspecting the page structure
                songs = page.query_selector_all('[class*="item"], [class*="card"], [class*="row"]')

                print(f"Found {len(songs)} potential song elements")

                for i, song in enumerate(songs[:max_songs]):
                    try:
                        # Extract text content
                        text_content = song.inner_text()

                        # Basic data structure
                        song_data = {
                            "rank": i + 1,
                            "scraped_at": datetime.now().isoformat(),
                            "raw_text": text_content,
                            # Additional fields will be populated when we analyze the structure
                            "title": "Unknown",
                            "artist": "Unknown",
                            "trend": "N/A"
                        }

                        songs_data.append(song_data)

                    except Exception as e:
                        print(f"⚠️  Error extracting song {i}: {e}")
                        continue

            except Exception as e:
                print(f"❌ Error during scraping: {e}")
                print("💡 Tip: The page structure may have changed. Check the screenshot and HTML.")

            finally:
                # Save raw HTML for analysis
                html_path = f"{self.data_dir}/page_source.html"
                with open(html_path, 'w', encoding='utf-8') as f:
                    f.write(page.content())
                print(f"💾 HTML source saved to {html_path}")

                browser.close()

        print(f"✅ Scraped {len(songs_data)} songs")
        return songs_data

    def save_json(self, data, filename="trending_music.json"):
        """Save data to JSON file"""
        filepath = f"{self.data_dir}/{filename}"
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        print(f"💾 JSON saved to {filepath}")

    def save_csv(self, data, filename="trending_music.csv"):
        """Save data to CSV file"""
        if not data:
            print("⚠️  No data to save to CSV")
            return

        filepath = f"{self.data_dir}/{filename}"

        # Get all unique keys from all dictionaries
        keys = set()
        for item in data:
            keys.update(item.keys())

        with open(filepath, 'w', newline='', encoding='utf-8') as f:
            writer = csv.DictWriter(f, fieldnames=sorted(keys))
            writer.writeheader()
            writer.writerows(data)

        print(f"📊 CSV saved to {filepath}")

    def run(self, max_songs=50):
        """Run the complete scraping workflow"""
        print("\n" + "="*60)
        print("🎵 TikTok Creative Center Music Scraper")
        print("="*60 + "\n")

        # Scrape data
        songs = self.scrape_trending_music(max_songs)

        if songs:
            # Save to both formats
            self.save_json(songs)
            self.save_csv(songs)

            # Display summary
            print("\n" + "="*60)
            print("📈 Summary")
            print("="*60)
            print(f"Total songs scraped: {len(songs)}")
            print(f"Data directory: {os.path.abspath(self.data_dir)}")
            print("\n✨ Scraping completed successfully!")
        else:
            print("\n⚠️  No data was scraped. Please check the screenshot and HTML source.")
            print("💡 You may need to update the selectors in the scraper.")

        return songs


def main():
    scraper = TikTokMusicScraper()
    scraper.run(max_songs=50)


if __name__ == "__main__":
    main()
